ts/src/base/Exchange.ts

```diff
@@ -3984,14 +3984,24 @@ export default class Exchange {
                 average = Precise.stringDiv (Precise.stringAdd (last, open), '2', precision);
             }
         }
-        if ((percentage === undefined) && (change !== undefined) && (open !== undefined) && Precise.stringGt (open, '0')) {
+        if ((percentage === undefined) && (change !== undefined) && (open !== undefined)) {
             percentage = Precise.stringMul (Precise.stringDiv (change, open), '100');
         }
         if ((change === undefined) && (percentage !== undefined) && (open !== undefined)) {
             change = Precise.stringDiv (Precise.stringMul (percentage, open), '100');
         }
-        if ((open === undefined) && (last !== undefined) && (change !== undefined)) {
-            open = Precise.stringSub (last, change);
+        if ((open === undefined) && (last !== undefined)) {
+            if (change !== undefined) {
+                open = Precise.stringSub (last, change);
+                if (percentage === undefined) {
+                    percentage = Precise.stringMul (Precise.stringDiv (change, open), '100');
+                }
+            } else if (percentage !== undefined) {
+                open = Precise.stringDiv (last, Precise.stringAdd ('1', Precise.stringDiv (percentage, '100')));
+                if (change === undefined) {
+                    change = Precise.stringDiv (Precise.stringMul (percentage, open), '100');
+                }
```

Comment:
@ttodua can you please specifiy what was wrong here?

Reply:
sure. look closely at current code: https://github.com/ccxt/ccxt/blob/master/ts/src/base/Exchange.ts#L3987-L3995

so, to follow, there we:
A) calc `percentage` from `open & change`
B) calc `change` from `open & percentage`
C) calc `open` from `close & change`

the below case was missing:
D) calc `open` from `close & percentage`

Reply:
so to explain further, imagine following data which passes into safeTicker: 
```
{
  last: 1.2,
  open: undefined,
  percentage: 20
}
```
with current master, neither `open` neither `change` is being calculated.

// if you passed `change` instead of `percentage` in above json, then `open` is being calculated, but not `percentage`

Reply:
so, we have to auto-calculate `change` (from `percentage`) or `percentage` (from `change`) **again**  after `open` price is auto-calculated.
atm, the `change & percentage` were being autocalculated only before `open` calculation (their clauses include `if (open !== undefined)`), but we have to repeat them again after `open` is calculated

Reply:
I think we should add a "base/statoc" test for this method like test.deepExtend where we ensure these derived fileds are correct

Reply:
> I think we should add a "base/statoc" test for this method like test.deepExtend where we ensure these derived fileds are correct

I've pushed quite recoded method, please review.

i further revealed  probably deeper issues in other "safe" methods too (eg. safeOrder, safeTrade, etc... need revisions), because auto calculation of fields does not work, if they are not repeated again. the `safeTickerCalculationHelper` shows there, that it needs to be called twice in `safeTicker`.